isotope/runner/istio.py

```diff
@@ -0,0 +1,192 @@
+
+    Requires Helm client to be present.
+
+    This downloads and extracts the archive in a temporary directory, then
+    installs the resources via `helm template` and `kubectl apply`.
+    """
+    with tempfile.TemporaryDirectory() as tmp_dir_path:
+        archive_path = os.path.join(tmp_dir_path, 'istio.tar.gz')
+        _download(archive_url, archive_path)
+
+        extracted_dir_path = os.path.join(tmp_dir_path, 'istio')
+        extracted_istio_path = _extract(archive_path, extracted_dir_path)
+
+        chart_path = os.path.join(extracted_istio_path, 'install',
+                                  'kubernetes', 'helm', 'istio')
+        _install(
+            chart_path,
+            consts.ISTIO_NAMESPACE,
+            intermediate_file_path=resources.ISTIO_GEN_YAML_PATH)
+
+        _create_ingress_rules(entrypoint_service_name,
+                              entrypoint_service_namespace)
+
+
+def get_ingress_gateway_url() -> str:
+    ip = wait.until_output([
+        'kubectl', '--namespace', consts.ISTIO_NAMESPACE, 'get', 'service',
+        'istio-ingressgateway', '-o',
+        'jsonpath={.status.loadBalancer.ingress[0].ip}'
+    ])
+    return 'http://{}:{}'.format(ip, consts.ISTIO_INGRESS_GATEWAY_PORT)
+
+
+# Adapted from
+# http://docs.python-requests.org/en/latest/user/advanced/#body-content-workflow.
+def _download(archive_url: str, path: str) -> None:
+    logging.info('downloading %s', archive_url)
+    with requests.get(archive_url, stream=True) as response:
+        with open(path, 'wb') as f:
+            for chunk in response.iter_content(chunk_size=1024):
+                if chunk:  # Filter out keep-alive new chunks.
+                    f.write(chunk)
+
+
+def _extract(archive_path: str, extracted_dir_path: str) -> str:
+    """Extracts the .tar.gz at archive_path to extracted_dir_path.
+
+    Args:
+        archive_path: path to a .tar.gz archive file, containing a single
+                directory when extracted
+        extracted_dir_path: the destination in which to extract the contents
+                of the archive
+
+    Returns:
+        the path to the single directory the archive contains
+    """
+    with tarfile.open(archive_path) as tar:
+        tar.extractall(path=extracted_dir_path)
+    extracted_items = os.listdir(extracted_dir_path)
+    if len(extracted_items) != 1:
+        raise ValueError(
+            'archive at {} did not contain a single directory'.format(
+                archive_path))
+    return os.path.join(extracted_dir_path, extracted_items[0])
+
+
+def _install(chart_path: str, namespace: str,
+             intermediate_file_path: str) -> None:
+    logging.info('installing Helm chart for Istio')
+    sh.run_kubectl(['create', 'namespace', namespace])
+    # TODO: Why is it necessary to set the hub and tag when these are already
+    # in the chart?
+    istio_yaml = sh.run(
+        [
+            'helm', 'template', chart_path, '--namespace', namespace,
```

Comment:
Right. I've thought about changing the set-up to use a list of values files rather than preset environments like "NONE" and "ISTIO". I'm just not sure the best way to represent "NONE" / "no istio" in that design.

I suggest we create an issue after this PR and we can flesh this out later?

Reply:
I also added a TODO.